numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -0,0 +1,432 @@
+}
+
+
+/*
+ * Core implementation of ndarray.__array_function__. This is exposed
+ * separately so we can avoid the overhead of a Python method call from
+ * within `implements_array_function`.
+ */
+NPY_NO_EXPORT PyObject *
+array_function_method_impl(PyObject *func, PyObject *types, PyObject *args,
+                           PyObject *kwargs)
+{
+    Py_ssize_t j;
+    PyObject *implementation, *result;
+
+    PyObject **items = PySequence_Fast_ITEMS(types);
+    Py_ssize_t length = PySequence_Fast_GET_SIZE(types);
+
+    for (j = 0; j < length; j++) {
+        PyObject *array_function;
+
+        int is_subclass = PyObject_IsSubclass(
+            items[j], (PyObject *)&PyArray_Type);
+        if (is_subclass == -1) {
+            return NULL;
+        }
+        if (is_subclass) {
+            continue;
+        }
+
+        array_function = PyArray_LookupSpecialFromType(
+            items[j], "__array_function__");
+        if (array_function == NULL) {
+            continue;
+        }
+        Py_DECREF(array_function);
+
+        Py_INCREF(Py_NotImplemented);
+        return Py_NotImplemented;
+    }
+
+    implementation = PyObject_GetAttrString(func, "__wrapped__");
+    if (implementation == NULL) {
+        return NULL;
+    }
+    result = PyObject_Call(implementation, args, kwargs);
+    Py_DECREF(implementation);
+    return result;
+}
+
+/*
+ * Calls __array_function__ on the provided argument, with a fast-path for
+ * ndarray.
+ */
+static PyObject *
+call_array_function(PyObject* argument, PyObject* method,
+                    PyObject* public_api, PyObject* types,
+                    PyObject* args, PyObject* kwargs)
+{
+    if (is_default_array_function(method)) {
+        return array_function_method_impl(public_api, types, args, kwargs);
+    }
+    else {
+        return PyObject_CallFunction(
+            method, "OOOOO", argument, public_api, types, args, kwargs);
+    }
+}
+
+/*
+ * Implements the __array_function__ protocol for a function, as described in
+ * in NEP-18.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    PyObject *types = NULL;
+    PyObject *implementing_args[NPY_MAXARGS];
+    PyObject *array_function_methods[NPY_MAXARGS];
+
+    int j;
+    int num_implementing_args = 0;
+    PyObject *result = NULL;
+
+    static PyObject *errmsg_formatter = NULL;
+
+    if (!PyArg_ParseTuple(
+            positional_args, "OOOOO:implement_array_function",
+            &implementation, &public_api, &relevant_args, &args, &kwargs)) {
+        return NULL;
+    }
+
+    relevant_args = PySequence_Fast(
+        relevant_args,
+        "dispatcher for __array_function__ did not return an iterable");
+    if (relevant_args == NULL) {
+        return NULL;
+    }
+
+    /* Fast path for the typical case of no-overrides. */
```

Comment:
I think it may be worth doing a bit of a speed check with just doing `get_implementing_args_and_methods` and checking for `num_implementing_args == 0 || num_implementing_args == 1 && <method==ndarray_array_function_>`. Accessing (or trying to access) attributes is quite expensive...

Reply:
I think the speed advantage might go the other way for longer lists of arguments (e.g., `concatenate` with a fairly long list of instances of the same class), though those again are not the most common case, and are relatively slow anyway.

But fine to leave as is with a comment for now!

Reply:
Actually, I think I was wrong. Here are the numbers I see when I measure this again on top of the latest changes. First, here's my patch:
```C
found_override = 0;
for (j = 0; j < num_implementing_args; j++) {
    if (!is_default_array_function(array_function_methods[j])) {
        found_override = 1;
        break;
    }
}
if (!found_override) {
    result = PyObject_Call(implementation, args, kwargs);
    goto cleanup;
}
```

Currently benchmarks:
```
bench_overrides.ArrayFunction.time_mock_broadcast_to_duck    1.92±0.01μs
bench_overrides.ArrayFunction.time_mock_broadcast_to_numpy   1.25±0.01μs
bench_overrides.ArrayFunction.time_mock_concatenate_duck     2.07±0.02μs
bench_overrides.ArrayFunction.time_mock_concatenate_many     2.60±0.08μs
bench_overrides.ArrayFunction.time_mock_concatenate_mixed    2.12±0.02μs
bench_overrides.ArrayFunction.time_mock_concatenate_numpy    1.38±0.02μs
```
With this patch:
```
bench_overrides.ArrayFunction.time_mock_broadcast_to_duck    1.63±0.02μs
bench_overrides.ArrayFunction.time_mock_broadcast_to_numpy   1.27±0.01μs
bench_overrides.ArrayFunction.time_mock_concatenate_duck     1.71±0.01μs
bench_overrides.ArrayFunction.time_mock_concatenate_many     3.21±0.06μs
bench_overrides.ArrayFunction.time_mock_concatenate_mixed    1.83±0.02μs
bench_overrides.ArrayFunction.time_mock_concatenate_numpy    1.40±0.01μs
```

Observations:
- Duck arrays operation have about 300 ns (15%) less overhead, which I guess is the speed of a CPython dict lookup.
- Pure numpy array operations are maybe slightly slower (20 ns?) but the difference is barely measurable, basically in the noise.
- The overhead for concatenating 1000 numpy arrays together *does* get noticeably slower, by about 600 ns. But again, this is basically unnoticeable compared to the entire concatenate operation (about 160 us).